flink-table/flink-table-planner/src/main/java/org/apache/flink/table/planner/plan/rules/logical/PushWatermarkIntoTableSourceScanRuleBase.java

```diff
@@ -117,8 +130,37 @@ protected FlinkLogicalTableSourceScan getNewScan(
                 idleTimeoutMillis = -1L;
             }
 
+            Optional<RelHint> optionsHintOptional =
+                    scan.getHints().stream()
+                            .filter(
+                                    relHint ->
+                                            relHint.hintName.equalsIgnoreCase(
+                                                    FlinkHints.HINT_NAME_OPTIONS))
+                            .findFirst();
+            Configuration hintOptions = new Configuration();
+            if (optionsHintOptional.isPresent()) {
+                RelHint optionsHint = optionsHintOptional.get();
+                hintOptions = Configuration.fromMap(optionsHint.kvOptions);
+            }
+            Configuration tableOptions = new Configuration();
+            RelOptTable table = scan.getTable();
+            if (table instanceof TableSourceTable) {
+                Map<String, String> tableConfigs =
+                        ((TableSourceTable) table)
+                                .contextResolvedTable()
+                                .getResolvedTable()
+                                .getOptions();
+                tableOptions = Configuration.fromMap(tableConfigs);
+            }
+            Optional<WatermarkParams> watermarkParamsOptional =
+                    parseWatermarkParams(hintOptions, tableOptions);
+
             final WatermarkPushDownSpec watermarkPushDownSpec =
-                    new WatermarkPushDownSpec(watermarkExpr, idleTimeoutMillis, producedType);
+                    new WatermarkPushDownSpec(
+                            watermarkExpr,
+                            idleTimeoutMillis,
+                            producedType,
+                            watermarkParamsOptional.orElse(null));
```

Comment:
I think the `watermarkParamsOptional` would not be empty. If so, why we still need to introduce `parseWatermarkParams` to generate the Optional parameter?

Moreover, if we refactor this, the `watermarkParams` in `WatermarkPushDownSpec` cannot be nullable.

Reply:
I hive modified the method `parseWatermarkParams` to return `WatermarkParams`.

However, the `watermarkParams` in  `WatermarkPushDownSpec ` must be nullable for compatibility.